app/assets/javascripts/application.js.erb

```diff
@@ -576,13 +551,67 @@ onPageLoad(() => {
     }
   });
 
+  on('click', 'button.comment-cancel', (event) => {
+    const comment = (parentSelector(event.target, '.comment'));
+    const commentId = comment.getAttribute('data-shortid');
+    if (commentId !== null && commentId !== '') {
+      fetch('/comments/' + commentId + '?show_tree_lines=true')
+        .then(response => {
+          response.text().then(text => replace(comment, text));
+        });
+    } else {
+      comment.remove();
+    }
+  });
+
   on('click', 'a.comment_editor', (event) => {
     let comment = parentSelector(event.target, '.comment');
-    fetch('/comments/' + comment.getAttribute('data-shortid') + '/edit')
+    const commentId = comment.getAttribute('data-shortid')
+    fetch('/comments/' + commentId + '/edit')
       .then(response => {
-        return response.text().then(function(text) {
-          replace(comment, text);
+        response.text().then(text => replace(comment, text));
+      });  
+    removeExtraInput(comment); 
+  });
+
+  on("click", "a.comment_deletor", (event) => {
+    event.preventDefault();
+    if (confirm("Are you sure you want to delete this comment?")) {
+      const comment = parentSelector(event.target, '.comment');
+      removeExtraInput(comment);
+      const commentId = comment.getAttribute('data-shortid');
+      fetchWithCSRF('/comments/' + commentId + '/delete',{method: 'post'})
+        .then(response => {
+          response.text().then(text => replace(comment, text));
+        });
+    }
+  });
+
+  on('click', 'a.comment_undeletor', (event) => {
+    event.preventDefault();
+    if (confirm("Are uou sure you want to undelete this comment?")) {
+      const comment = parentSelector(event.target, '.comment');
+      removeExtraInput(comment);
+      const commentId = comment.getAttribute('data-shortid');
+      fetchWithCSRF('/comments/' + commentId + '/undelete', {method: 'post'})
+        .then(response => {
+          response.text().then(text => replace(comment, text));
         });
+    }
+  });
+
+  on('click', 'a.comment_moderator', (event) => {
+    const reason = prompt("Moderation reason:");
+    if (reason == null || reason == '') 
+      return false;
+
+    const formData = new FormData();
+    formData.append('reason', reason);
+    const comment = parentSelector(event.target, '.comment');
+    const commentId = comment.getAttribute('data-shortid');
+    fetchWithCSRF('/comments/' + commentId + '/delete', { method: 'post', body: formData })
```

Comment:
I think this needs a `removeExtraInputs` call as well?

Reply:
This got covered when I moved the `removeExtraInputs` to the replace function.